File: v1/src/main/java/com/api/v1/utils/PaymentFinderUtil.java
package com.api.v1.utils;

import com.api.v1.domain.Payment;
import com.api.v1.domain.PaymentRepository;
import com.api.v1.exceptions.PaymentNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.math.BigInteger;

@Component
public class PaymentFinderUtil {

    @Autowired
    private PaymentRepository repository;

    public Mono<Payment> find(String orderNumber) {
        return repository
                .findAll()
                .filter(e -> e.getOrderNumber().equals(new BigInteger(orderNumber)))
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new PaymentNotFoundException(orderNumber)));
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/payment/utils/PaymentFinderUtil.kt
package com.api.v2.payment.utils

import com.api.v2.payment.domain.Payment
import com.api.v2.payment.domain.PaymentRepository
import com.api.v2.payment.exceptions.PaymentNotFoundException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.singleOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.math.BigInteger

@Component
class PaymentFinderUtil {

    @Autowired
    private lateinit var repository: PaymentRepository

    suspend fun find(orderNumber: String): Payment {
        return withContext(Dispatchers.IO) {
            repository
                .findAll()
                .filter { e -> e.orderNumber == BigInteger(orderNumber) }
                .singleOrNull() ?: throw PaymentNotFoundException(BigInteger(orderNumber))
        }
    }

}